pkg/tracing/tracing.go

```diff
@@ -0,0 +1,164 @@
+
+	AttrGenAIRequestModel = "gen_ai.request.model"
+	AttrOperationOutcome  = "operation.outcome"
+
+	OutcomeSuccess = "success"
+	OutcomeError   = "error"
+)
+
+type Config struct {
+	Enabled          bool
+	ExporterEndpoint string
+	SamplingRate     float64
+	ServiceName      string
+}
+
+func NewConfigFromEnv() *Config {
+	config := &Config{
+		Enabled:          false,
+		ExporterEndpoint: "http://localhost:4317",
+		SamplingRate:     0.1,
+		ServiceName:      ServiceName,
+	}
+
+	if enabled := os.Getenv(envOTELTracingEnabled); enabled != "" {
+		if enabledBool, err := strconv.ParseBool(enabled); err == nil {
+			config.Enabled = enabledBool
+		}
+	}
+
+	if endpoint := os.Getenv(envOTELExporterEndpoint); endpoint != "" {
+		config.ExporterEndpoint = endpoint
+	}
+
+	if serviceName := os.Getenv(envOTELServiceName); serviceName != "" {
+		config.ServiceName = serviceName
+	}
+
+	if samplingRate := os.Getenv(envOTELSamplingRate); samplingRate != "" {
+		if rate, err := strconv.ParseFloat(samplingRate, 64); err == nil {
+			config.SamplingRate = rate
+		}
+	}
+
+	return config
+}
+
+// Initialize sets up OpenTelemetry tracing with the given configuration.
+// It always sets up context propagation, even if tracing is disabled.
+func Initialize(ctx context.Context, config *Config) (func(), error) {
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+
+	if !config.Enabled {
+		// Return a no-op shutdown function if tracing is disabled
+		return func() {}, nil
+	}
+
+	conn, err := grpc.DialContext(ctx, config.ExporterEndpoint,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+		grpc.WithTimeout(5*time.Second),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gRPC connection to collector: %w", err)
+	}
+
+	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create OTLP trace exporter: %w", err)
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(config.ServiceName),
+			semconv.ServiceVersionKey.String("1.0.0"),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exporter),
+		sdktrace.WithResource(res),
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(config.SamplingRate)),
+	)
+
+	otel.SetTracerProvider(tp)
+
+	return func() {
+		tp.Shutdown(context.Background())
+	}, nil
+}
+
+func StartSpan(ctx context.Context, name, operation string) (context.Context, trace.Span) {
+	tracer := otel.Tracer(ServiceName)
+	return tracer.Start(ctx, name)
+}
+
+func StartGatewaySpan(ctx context.Context, operation string) (context.Context, trace.Span) {
+	ctx, span := StartSpan(ctx, operation, operation)
+	// TODO: Add common gateway attributes here
+	return ctx, span
+}
```

Comment:
Maybe we want to replace them with something like:
```golang
func SetGatewayAttributes(span trace.Span, gateway Gateway) {
	if gateway == nil {
		return
	}

	span.SetAttributes(
		AttrGatewaySomething.String(gateway.Something()),
	)
}

```
Allowing a user to set attributes without the need to know the specific Keys?